micro_http/src/connection.rs

```diff
@@ -0,0 +1,634 @@
+                        self.body_bytes_to_be_read = 0;
+                    }
+
+                    let request = self.request.as_mut().unwrap();
+                    // If there are no more bytes to be read for this request.
+                    if self.body_bytes_to_be_read == 0 {
+                        // Assign the body of the request.
+                        let placeholder: Vec<_> = self
+                            .body_vec
+                            .drain(..request.headers.content_length() as usize)
+                            .collect();
+                        request.body = Some(Body::new(placeholder));
+
+                        // If we read more bytes than we should have into the body of the request.
+                        if !self.body_vec.is_empty() {
+                            return Err(ConnectionError::StreamError);
+                        }
+
+                        self.state = ConnectionState::RequestReady;
+                    }
+                }
+                ConnectionState::RequestReady => {
+                    // This request is ready to be passed for handling.
+                    // Update the state machine to expect a new request and push this request into
+                    // the `parsed_requests` queue.
+                    self.state = ConnectionState::WaitingForRequestLine;
+                    self.body_bytes_to_be_read = 0;
+                    self.parsed_requests.push_back(self.request.take().unwrap());
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Tries to write responses to the provided stream.
+    /// Meant to be used only with non-blocking streams and an `EPOLL` structure.
+    /// Should be called whenever an `EPOLLOUT` event is signaled.
+    pub fn try_write(&mut self) -> Result<(), ConnectionError> {
+        if self.response_buffer.is_none() {
+            if let Some(response) = self.response_queue.pop_front() {
+                let mut response_buffer_vec: Vec<u8> = Vec::new();
+                response
+                    .write_all(&mut response_buffer_vec)
+                    .map_err(|_| ConnectionError::StreamError)?;
+                self.response_buffer = Some(response_buffer_vec);
+            } else {
+                return Ok(());
+            }
+        }
+
+        let mut response_fully_written = false;
+
+        if let Some(response_buffer_vec) = self.response_buffer.as_mut() {
+            let bytes_to_be_written = response_buffer_vec.len();
+            match self.stream.write(response_buffer_vec.as_slice()) {
+                Ok(0) => {
+                    return Err(ConnectionError::ConnectionClosed);
+                }
+                Ok(bytes_written) => {
+                    if bytes_written != bytes_to_be_written {
+                        response_buffer_vec.drain(..bytes_written);
+                    } else {
+                        response_fully_written = true;
```

Comment:
Completely drop `response_fully_written` and just do `response_buffer_vec.take()` here.

Reply:
That was my initial idea, but the rust borrowing rules don't let me do that as it would be a second mutable borrow.